source/solvers/vof.cc

```diff
@@ -1082,5 +1116,219 @@ VolumeOfFluid<dim>::assemble_mass_matrix_diagonal(
+  const unsigned int peeling_threshold =
+    this->simulation_parameters.boundary_conditions_vof.peeling_threshold[i_bc];
+
+  // Loop on cell_vof
+  for (const auto &cell_vof : dof_handler.active_cell_iterators())
+    {
+      if (cell_vof->is_locally_owned() && cell_vof->at_boundary())
+        {
+          for (const auto face : cell_vof->face_indices())
+            {
+              if (cell_vof->face(face)->at_boundary() &&
+                  cell_vof->face(face)->boundary_id() == boundary_id)
+                {
+                  // Equivalent dof_handler for CFD
+                  typename DoFHandler<dim>::active_cell_iterator cell_cfd(
+                    &(*(this->triangulation)),
+                    cell_vof->level(),
+                    cell_vof->index(),
+                    dof_handler_cfd);
+
+                  // Reinit fe_face_values for CFD and VOF
+                  fe_face_values_vof.reinit(cell_vof, face);
+                  fe_face_values_cfd.reinit(cell_cfd, face);
+
+                  // Get pressure values
+                  fe_face_values_cfd[pressure].get_function_values(
+                    current_solution_cfd, pressure_values);
+
+                  // Get phase values
+                  fe_face_values_vof.get_function_values(present_solution,
+                                                         phase_values);
+
+                  // Number of points meeting the peeling criteria in the cell
+                  unsigned int nb_q_peeled(0);
+
+                  // Number of points meeting the wetting criteria in the cell
+                  unsigned int nb_q_wet(0);
+
+                  // Local index (see deal.II step 4)
+                  cell_vof->get_dof_indices(dof_indices_vof);
+
+                  for (unsigned int q = 0; q < n_q_points; q++)
+                    {
+                      if (pressure_values[q] > wetting_threshold)
+                        {
+                          // wetting of lowest density fluid
+                          if ((density_fluid_1 > density_fluid_0) &&
+                              (phase_values[q] < 0.5))
+                            {
+                              // increment cell count
+                              nb_q_wet++;
+
+                              // if enough quadrature points meet the condition
+                              if (nb_q_wet > n_q_points / 2)
+                                {
+                                  change_cell_phase("w",
+                                                    1,
+                                                    solution_pw,
+                                                    dof_indices_vof);
+                                }
+                            }
+                          else if ((density_fluid_0 > density_fluid_1) &&
+                                   (phase_values[q] > 0.5))
+                            {
+                              // increment cell count
```

Comment:
This is a quadrature point count no? Why cell count?

Reply:
yes you are right, I will correct that